src/cpio/client_providers/kms_client_provider/azure/azure_kms_client_provider.cc

```diff
@@ -0,0 +1,377 @@
+                      execution_result, "Failed to get the session token.");
+    decrypt_context.Finish(execution_result);
+
+    return absl::UnknownError(google::scp::core::errors::GetErrorMessage(
+        execution_result.status_code));
+  }
+
+  return absl::OkStatus();
+}
+
+void AzureKmsClientProvider::GetSessionCredentialsCallbackToDecrypt(
+    core::AsyncContext<DecryptRequest, DecryptResponse>& decrypt_context,
+    core::AsyncContext<GetSessionTokenRequest, GetSessionTokenResponse>&
+        get_token_context) noexcept {
+  if (!get_token_context.result.Successful()) {
+    SCP_ERROR_CONTEXT(kAzureKmsClientProvider, decrypt_context,
+                      get_token_context.result,
+                      "Failed to get the access token.");
+    decrypt_context.result = get_token_context.result;
+    decrypt_context.Finish();
+    return;
+  }
+
+  const auto& access_token = *get_token_context.response->session_token;
+
+  const auto& ciphertext = decrypt_context.request->ciphertext();
+  if (ciphertext.empty()) {
+    auto execution_result = FailureExecutionResult(
+        SC_AZURE_KMS_CLIENT_PROVIDER_CIPHER_TEXT_NOT_FOUND);
+    SCP_ERROR_CONTEXT(kAzureKmsClientProvider, decrypt_context,
+                      execution_result,
+                      "Failed to get cipher text from decryption request.");
+    decrypt_context.result = execution_result;
+    decrypt_context.Finish();
+    return;
+  }
+
+  // Check that there is an ID for the key to decrypt with
+  const auto& key_id = decrypt_context.request->key_resource_name();
+  if (key_id.empty()) {
+    auto execution_result =
+        FailureExecutionResult(SC_AZURE_KMS_CLIENT_PROVIDER_KEY_ID_NOT_FOUND);
+    SCP_ERROR_CONTEXT(kAzureKmsClientProvider, decrypt_context,
+                      execution_result,
+                      "Failed to get Key ID from decryption request.");
+    decrypt_context.result = execution_result;
+    decrypt_context.Finish();
+    return;
+  }
+
+  AsyncContext<HttpRequest, HttpResponse> http_context;
+  http_context.request = std::make_shared<HttpRequest>();
+
+  // For the first call, it tries to get the unwrap URL from environment
+  // variable. This is done here because Init() is not called by the shared code
+  // and it's a temporary workaround.
+  if (unwrap_url_.empty()) {
+    const char* value_from_env = std::getenv(kAzureKmsUnwrapUrlEnvVar);
+    if (value_from_env) {
+      unwrap_url_ = value_from_env;
+    } else {
+      unwrap_url_ = kDefaultKmsUnwrapPath;
+    }
+  }
+  http_context.request->path = std::make_shared<Uri>(unwrap_url_);
+  http_context.request->method = HttpMethod::POST;
+
+  std::shared_ptr<EvpPkeyWrapper> public_key;
+  std::shared_ptr<EvpPkeyWrapper> private_key;
+
+  // Temporary store wrapping_key
+  std::pair<std::shared_ptr<EvpPkeyWrapper>, std::shared_ptr<EvpPkeyWrapper>>
+      wrapping_key_pair;
+  std::string hex_hash_on_wrapping_key = "";
+  if (hasSnp()) {
```

Comment:
I have not seen this pattern in google code.

Reply:
Yeah, you are correct, but we decided to do this way while knowing the upstream has different .cc code for tee and non-tee. I remember it's so that we can use the same binary everywhere.